### PR TITLE
Faster ref-count logic for when ref-counted object does not escape

### PR DESCRIPTION
**Pull request by original-brownbear**
Introducing a plain version of `AbstractRefCounted` as a compromise. This saves a bunch of allocations and a circular reference to the object holding the ref counted instance, making smaller SearchHit instances etc. cheaper. We could get an even more direct solution here by making these extend `AbstractRefCounted` but that would lose us the ability to leak-track in tests, so doing it this way (same way Netty does it on their end) as a compromise.

### CODE REVIEW

File: libs/core/src/main/java/org/elasticsearch/core/AbstractRefCounted.java [hunk 1]
```diff
@@ -125,4 +125,15 @@ public String toString() {
         };
     }
 
+    /**
+     * Constructs an {@link AbstractRefCounted} which does nothing when all references are released. It is the responsibility of the caller
+     * to run whatever release logic should be executed when {@link AbstractRefCounted#decRef()} returns true.
+     */
+    public static RefCounted plain() {
+        return new AbstractRefCounted() {
```
Comment (thecoop) [inline libs/core/src/main/java/org/elasticsearch/core/AbstractRefCounted.java:133]: Given how many static methods are on `AbstractRefCounted`, and its referenced directly, is it worth renaming the class to something like `DefaultRefCounted`, `SimpleRefCounted`, `BaseRefCounted` etc?

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by DaveCTurner** [comment]
Oh one question:

> for when ref-counted object does not escape

What does this mean exactly? I'm not sure what you mean by "escape", and it seems like this pattern should work anywhere.

**Comment by original-brownbear** [comment]
> What does this mean exactly? I'm not sure what you mean by "escape", and it seems like this pattern should work anywhere.

I was thinking about situation where we take the `RefCounted` of one instance and assign it to another instance (as we do when slicing `ReleasableBytesReference`. In that case all the slices reference the original Netty buffer wrapped in the ref-counted instance and we can't easily figure out what we need to release on a true return from `decRef`. We could address this by passing the original buffer as the thing to release to the slices or so (kinda what Netty's `ByteBuf` do) but I'm not yet convinced there's performance benefit to it. I'll try tricks like that shortly the step in https://elasticsearch-benchmarks.elastic.co/index.html#tracks/tsdb/nightly/default/90d (nightly-tsdb-standard-query-default_1k-latency) just isn't going way :P but lets see how far this gets us first :)

**Comment by DaveCTurner** [comment]
Right gotcha I see, this only works if there's one path to calling `decRef()` indeed.

**Comment by original-brownbear** [comment]
Thanks both! Renamed the class and moved it to the top level. => merging

**Pull request merged**